index.js

```diff
@@ -53,49 +54,55 @@ var authS3O = function(req, res, next) {
 
 	if (req.cookies === undefined || req.cookies === null) {
 		var cookies = req.headers.cookie;
-	    if (cookies) {
+		if (cookies) {
 			req.cookies = cookieParser(cookies);
-	    } else {
+		} else {
 			req.cookies = Object.create(null);
-	    }
+		}
 	}
 
 	// Check for s3o username/token URL parameters.
 	// These parameters come from https://s3o.ft.com. It redirects back after it does the google authentication.
-	if (req.query.username && req.query.token) {
-		debug("S3O: Found parameter token for s3o_username: " + req.query.username);
-
-		if (authenticateToken(res, req.query.username, req.query.token)) {
-
-			// Strip the username and token from the URL (but keep any other parameters)
-			delete req.query['username'];
-			delete req.query['token'];
-			var cleanURL = url.parse(req.path);
-			cleanURL.query = req.query;
-
-			debug("S3O: Parameters detected in URL. Redirecting to base path: " + url.format(cleanURL));
-			debug("S3O: " + JSON.stringify(req.path));
-
-			// Don't cache any redirection responses.
-			res.header("Cache-Control", "private, no-cache, no-store, must-revalidate");
-			res.header("Pragma", "no-cache");
-			res.header("Expires", 0);
-			res.redirect(url.format(cleanURL));
-		}
+	if (req.method === 'POST' && req.query.username) {
+		urlencoded(req, res, function() {
+				debug("S3O: Found parameter token for s3o_username: " + req.query.username);
+
+				if (authenticateToken(res, req.query.username, req.hostname, req.body.token)) {
+
+					// Strip the username and token from the URL (but keep any other parameters)
+					delete req.query['username'];
+					delete req.query['token'];
+					var cleanURL = url.parse(req.path);
+					cleanURL.query = req.query;
+
+					debug("S3O: Parameters detected in URL and body. Redirecting to base path: " + url.format(cleanURL));
+					debug("S3O: " + JSON.stringify(req.path));
+
+					// Don't cache any redirection responses.
+					res.header("Cache-Control", "private, no-cache, no-store, must-revalidate");
+					res.header("Pragma", "no-cache");
+					res.header("Expires", 0);
+					res.redirect(url.format(cleanURL));
+				} else {
+					res.send("<h1>Authentication error.</h1><p>For access, please login with your FT account</p>");
+				}
+			});
 	}
 	// Check for s3o username/token cookies
 	else if (req.cookies.s3o_username && req.cookies.s3o_token) {
 		debug("S3O: Found cookie token for s3o_username: " + req.cookies.s3o_username);
 
-		if (authenticateToken(res, req.cookies.s3o_username, req.cookies.s3o_token)) {
+		if (authenticateToken(res, req.cookies.s3o_username, req.hostname, req.cookies.s3o_token)) {
 			next();
+		} else {
+			res.send("<h1>Authentication error.</h1><p>For access, please login with your FT account</p>");
```

Comment:
got a bit confused when the helper function actually sent the headers…  this way the helper function is a bit ‘safer’… in a later PR I plan to abstract the error sending into a single method that all the errors that this middleware generates can be output